test/resources/qemu-guest-agent.py

```
@@ -431,3 +434,129 @@ def guest_agent_is_ready(vm_name: str):
+            print("QEMU agent is now available")
+            return
+        except Exception:
+            print(f"Attempt {i}/{max_retries}: QEMU agent not ready, retrying in {sleep_time} seconds")
+            sleep(sleep_time)
+
+
+def main():
+    parser = argparse.ArgumentParser()
+
+    subparsers = parser.add_subparsers(dest="command", required=True)
+
+    cp_from_parser = subparsers.add_parser("download", help="Copy file(s) from VM")
+    cp_from_parser.add_argument("--vm", required=True, help="domain name")
+    cp_from_parser.add_argument("--src_dir", required=True, help="source path on VM")
+    cp_from_parser.add_argument("--dst_dir", required=True, help="local destination path")
+    cp_from_parser.add_argument("--pat", required=True, help="filename or pattern")
+
+    cp_to_parser = subparsers.add_parser("upload", help="Copy file to VM")
+    cp_to_parser.add_argument("--vm", required=True, help="domain name")
+    cp_to_parser.add_argument("--src", required=True, help="path to local file")
+    cp_to_parser.add_argument("--dst", required=True, help="destination path on VM")
+
+    bash_parser = subparsers.add_parser("bash", help="Run bash command on VM")
+    bash_parser.add_argument("--vm", required=True, help="domain name")
+    bash_parser.add_argument("--args", required=True, help="arguments")
+
+    wait_parser = subparsers.add_parser("wait", help="Wait for guest agent to be ready")
+    wait_parser.add_argument("--vm", required=True, help="domain name")
+
+    args = parser.parse_args()
+
+    if args.command == "download":
+        download_files(args.vm, args.src_dir, args.dst_dir, args.pat)
+    elif args.command == "upload":
+        upload_file(args.vm, args.src, args.dst)
+    elif args.command == "bash":
+        print(f"Running {args.args}")
+        run_guest_process(args.vm, "/bin/bash", "-c", args.args)
+    elif args.command == "wait":
+        wait_for_guest_agent(args.vm)
+
+
+if __name__ == "__main__":
+    main()
```

Do we need the `main` function at all? I mean, it does not do much except passing arguments to other functions. What about just calling those functions from bash directly?

---

I felt like that would make it ugly in the `scenario.sh` script, doing it this way makes it more readable.